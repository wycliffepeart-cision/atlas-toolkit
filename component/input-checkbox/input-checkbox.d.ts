export declare const CHECKBOX_ATTRIBUTES: {
    DISABLED: string;
    REQUIRED: string;
    CHECKED: string;
    NAME: string;
    VALUE: string;
};
export declare class InputCheckbox extends HTMLElement {
    #private;
    static get observedAttributes(): string[];
    attributeChangedCallback(name: any, _oldValue: any, newValue: any): void;
    connectedCallback(): void;
    disconnectedCallback(): void;
    get checked(): boolean;
    set checked(state: boolean);
    get value(): string;
    set value(value: string);
}
//# sourceMappingURL=input-checkbox.d.ts.map