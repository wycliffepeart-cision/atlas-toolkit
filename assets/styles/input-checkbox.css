@layer atlas {
	atlas-checkbox {
		--custom-accent-color: #0b9ecd;
		--custom-outline-color: #0e7799;
		--box-width: 1.35em;
		--gap: 0.5em;
	}

	atlas-checkbox {
		display: flex;
		align-items: center;
		gap: var(--gap);
	}

	atlas-checkbox::part(input) {
		cursor: pointer;
		width: var(--box-width);
		aspect-ratio: 1 / 1;

		accent-color: var(--atlas-input-common-accent-color);
		outline-color: var(--atlas-input-common-accent-color);
		outline-offset: -1px;
	}

	atlas-checkbox::part(input):focus-visible {
		box-shadow: 0 0 5px var(--atlas-input-common-accent-color);
	}

	atlas-checkbox::part(input):hover {
		accent-color: var(--custom-accent-color);
		outline-color: var(--custom-outline-color);
	}

	atlas-checkbox::part(input):active {
		outline: none;
		box-shadow: none;
	}

	atlas-checkbox::part(label) {
		cursor: pointer;
	}

	atlas-checkbox::part(required) {
		display: none;
	}

	atlas-checkbox[required]:not([required="false"])::part(required) {
		display: inline;
	}

	atlas-checkbox::part(helper-text) {
		display: none;
	}
}
