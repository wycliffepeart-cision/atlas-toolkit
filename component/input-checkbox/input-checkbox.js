import { _ as __decorate } from '../../tslib.es6-2e0713a3.js';
import { Template } from '../../decorators/template.js';

var InputCheckboxTemplate = "<template>\n  <input id=\"Input\" type=\"checkbox\" part=\"input\">\n\n  <label for=\"Input\" part=\"label-wrapper\">\n    <slot name=\"label\" part=\"label\"></slot>\n    <slot name=\"required\" part=\"required\">*</slot>\n  </label>\n\n  <slot name=\"helperText\" part=\"helper-text\"></slot>\n  <slot name=\"validationText\" part=\"validation-text\"></slot>\n</template>\n";

const CHECKBOX_ATTRIBUTES = {
    DISABLED: 'disabled',
    REQUIRED: 'required',
    CHECKED: 'checked',
    NAME: 'name',
    VALUE: 'value',
};
let InputCheckbox = class InputCheckbox extends HTMLElement {
    static get observedAttributes() {
        return Object.values(CHECKBOX_ATTRIBUTES);
    }
    attributeChangedCallback(name, _oldValue, newValue) {
        this.#attributePassthrought(name, newValue);
    }
    connectedCallback() {
        if (!this.isConnected) {
            return;
        }
        this.#inputElem.addEventListener('change', this.#handleInputChange);
    }
    disconnectedCallback() {
        this.#inputElem.removeEventListener('change', this.#handleInputChange);
    }
    #attributePassthrought(name, value, exclude = []) {
        if (exclude.includes(name)) {
            return;
        }
        const input = this.shadowRoot.getElementById('Input');
        if (value === null) {
            input.removeAttribute(name);
        }
        else {
            input.setAttribute(name, value);
        }
    }
    #handleInputChange = () => {
        this.dispatchEvent(new CustomEvent('change', { bubbles: true, composed: true }));
    };
    get #inputElem() {
        return this.shadowRoot.getElementById('Input');
    }
    get checked() {
        return this.#inputElem.checked;
    }
    set checked(state) {
        this.#inputElem.checked = state;
    }
    get value() {
        return this.#inputElem.value;
    }
    set value(value) {
        this.#inputElem.value = value;
    }
};
InputCheckbox = __decorate([
    Template(InputCheckboxTemplate)
], InputCheckbox);
if (!customElements.get('atlas-checkbox'))
    customElements.define('atlas-checkbox', InputCheckbox);

export { CHECKBOX_ATTRIBUTES, InputCheckbox };
