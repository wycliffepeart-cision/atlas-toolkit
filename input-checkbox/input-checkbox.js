import { _ as __decorate, a as __classPrivateFieldGet } from '../tslib.es6-806de115.js';
import { Template } from '../decorators/template.js';

var InputCheckboxTemplate = "<template>\n  <input id=\"Input\" type=\"checkbox\" part=\"input\">\n\n  <label for=\"Input\" part=\"label-wrapper\">\n    <slot name=\"label\" part=\"label\"></slot>\n    <slot name=\"required\" part=\"required\">*</slot>\n  </label>\n\n  <slot name=\"helperText\" part=\"helper-text\"></slot>\n  <slot name=\"validationText\" part=\"validation-text\"></slot>\n</template>\n";

var _InputCheckbox_instances, _InputCheckbox_attributePassthrought, _InputCheckbox_handleInputChange, _InputCheckbox_inputElem_get;
const CHECKBOX_ATTRIBUTES = {
    DISABLED: 'disabled',
    REQUIRED: 'required',
    CHECKED: 'checked',
    NAME: 'name',
    VALUE: 'value',
};
let InputCheckbox = class InputCheckbox extends HTMLElement {
    constructor() {
        super(...arguments);
        _InputCheckbox_instances.add(this);
        _InputCheckbox_handleInputChange.set(this, () => {
            this.dispatchEvent(new CustomEvent('change', { bubbles: true, composed: true }));
        });
    }
    static get observedAttributes() {
        return Object.values(CHECKBOX_ATTRIBUTES);
    }
    attributeChangedCallback(name, _oldValue, newValue) {
        __classPrivateFieldGet(this, _InputCheckbox_instances, "m", _InputCheckbox_attributePassthrought).call(this, name, newValue);
    }
    connectedCallback() {
        if (!this.isConnected) {
            return;
        }
        __classPrivateFieldGet(this, _InputCheckbox_instances, "a", _InputCheckbox_inputElem_get).addEventListener('change', __classPrivateFieldGet(this, _InputCheckbox_handleInputChange, "f"));
    }
    disconnectedCallback() {
        __classPrivateFieldGet(this, _InputCheckbox_instances, "a", _InputCheckbox_inputElem_get).removeEventListener('change', __classPrivateFieldGet(this, _InputCheckbox_handleInputChange, "f"));
    }
    get checked() {
        return __classPrivateFieldGet(this, _InputCheckbox_instances, "a", _InputCheckbox_inputElem_get).checked;
    }
    set checked(state) {
        __classPrivateFieldGet(this, _InputCheckbox_instances, "a", _InputCheckbox_inputElem_get).checked = state;
    }
    get value() {
        return __classPrivateFieldGet(this, _InputCheckbox_instances, "a", _InputCheckbox_inputElem_get).value;
    }
    set value(value) {
        __classPrivateFieldGet(this, _InputCheckbox_instances, "a", _InputCheckbox_inputElem_get).value = value;
    }
};
_InputCheckbox_handleInputChange = new WeakMap();
_InputCheckbox_instances = new WeakSet();
_InputCheckbox_attributePassthrought = function _InputCheckbox_attributePassthrought(name, value, exclude = []) {
    if (exclude.includes(name)) {
        return;
    }
    const input = this.shadowRoot.getElementById('Input');
    if (value === null) {
        input.removeAttribute(name);
    }
    else {
        input.setAttribute(name, value);
    }
};
_InputCheckbox_inputElem_get = function _InputCheckbox_inputElem_get() {
    return this.shadowRoot.getElementById('Input');
};
InputCheckbox = __decorate([
    Template(InputCheckboxTemplate)
], InputCheckbox);

export { CHECKBOX_ATTRIBUTES, InputCheckbox };
