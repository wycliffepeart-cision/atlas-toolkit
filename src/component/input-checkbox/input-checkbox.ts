import { Template } from '../../decorators/template';
import InputCheckboxTemplate from './input-checkbox.html';

export const CHECKBOX_ATTRIBUTES = {
  DISABLED: 'disabled',
  REQUIRED: 'required',
  CHECKED: 'checked',
  NAME: 'name',
  VALUE: 'value',
};

@Template(InputCheckboxTemplate)
export class InputCheckbox extends HTMLElement {
  static get observedAttributes() {
    return Object.values(CHECKBOX_ATTRIBUTES);
  }

  attributeChangedCallback(name, _oldValue, newValue) {
    this.#attributePassthrought(name, newValue);
  }

  connectedCallback() {
    if (!this.isConnected) {
      return;
    }

    this.#inputElem.addEventListener('change', this.#handleInputChange);
  }

  disconnectedCallback() {
    this.#inputElem.removeEventListener('change', this.#handleInputChange);
  }

  #attributePassthrought(name, value, exclude = []) {
    if (exclude.includes(name)) {
      return;
    }

    const input = this.shadowRoot.getElementById('Input');
    if (value === null) {
      input.removeAttribute(name);
    } else {
      input.setAttribute(name, value);
    }
  }

  #handleInputChange = () => {
    this.dispatchEvent(new CustomEvent('change', { bubbles: true, composed: true }));
  };

  get #inputElem(): HTMLInputElement {
    return this.shadowRoot.getElementById('Input') as unknown as HTMLInputElement;
  }

  get checked() {
    return this.#inputElem.checked;
  }

  set checked(state) {
    this.#inputElem.checked = state;
  }

  get value() {
    return this.#inputElem.value;
  }

  set value(value) {
    this.#inputElem.value = value;
  }
}
