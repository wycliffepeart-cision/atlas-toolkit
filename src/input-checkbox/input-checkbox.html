<template>
  <input id="Input" type="checkbox" part="input" />

  <label for="Input" part="label-wrapper">
    <slot name="label" part="label"></slot>
    <slot name="required" part="required">*</slot>
  </label>

  <slot name="helperText" part="helper-text"></slot>
  <slot name="validationText" part="validation-text"></slot>
</template>
