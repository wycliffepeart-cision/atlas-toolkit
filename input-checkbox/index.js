import { InputCheckbox } from './input-checkbox.js';
export { CHECKBOX_ATTRIBUTES } from './input-checkbox.js';
import '../tslib.es6-806de115.js';
import '../decorators/template.js';

if (!customElements.get('atlas-checkbox'))
    customElements.define('atlas-checkbox', InputCheckbox);
