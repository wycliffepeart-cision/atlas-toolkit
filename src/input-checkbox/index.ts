import { InputCheckbox } from './input-checkbox';
export { CHECKBOX_ATTRIBUTES } from './input-checkbox';

if (!customElements.get('atlas-checkbox')) customElements.define('atlas-checkbox', InputCheckbox);
