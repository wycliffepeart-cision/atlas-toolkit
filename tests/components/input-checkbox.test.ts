import { fireEvent } from '@testing-library/dom';

import { CHECKBOX_ATTRIBUTES } from '../../src/component/input-checkbox/input-checkbox';

const ELEMENT_TAG = 'atlas-checkbox';

afterEach(() => {
  const element = document.querySelector('body > *:first-child');
  element?.remove();
});

describe('InputCheckbox component', () => {
  it('constructs', () => {
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    (expect(element) as any).toBeInTheDocument();
  });

  it('match snapshot', () => {
    const element = document.createElement(ELEMENT_TAG);
    (expect(element.shadowRoot.innerHTML) as any).toMatchSnapshot();
  });

  it('triggers onChange proxy event when updated', () => {
    const element = document.createElement(ELEMENT_TAG);
    document.body.appendChild(element);

    const innerInput = element.shadowRoot.getElementById('Input');

    const mockChangeEvent = jest.fn();

    element.addEventListener('change', mockChangeEvent);
    fireEvent.click(innerInput);
    (expect(mockChangeEvent) as any).toHaveBeenCalledTimes(1);
  });

  it('passthrought observed attributes', () => {
    const element = document.createElement(ELEMENT_TAG);
    Object.values(CHECKBOX_ATTRIBUTES).forEach((attr) => {
      const { attributes } = element.shadowRoot.getElementById('Input');

      element.setAttribute(attr, attr);
      (expect(attributes[attr].value) as any).toBe(attr);

      element.removeAttribute(attr);
      (expect(attributes[attr]) as any).toBe(undefined);
    });
  });

  describe('JS-API', () => {
    describe('.checked', () => {
      it('returns false when default', () => {
        const element: any = document.createElement(ELEMENT_TAG);
        (expect(element.checked) as any).toBe(false);
      });

      it('returns true when set to true', () => {
        const element: any = document.createElement(ELEMENT_TAG);
        element.checked = true;
        (expect(element.checked) as any).toBe(true);
      });
    });

    describe('.value', () => {
      it('returns default checkbox value', () => {
        const element: any = document.createElement(ELEMENT_TAG);
        (expect(element.value) as any).toBe('on');
      });

      it('returns custom value when set', () => {
        const element: any = document.createElement(ELEMENT_TAG);
        element.value = 'custom value';
        (expect(element.value) as any).toBe('custom value');
      });
    });
  });
});
