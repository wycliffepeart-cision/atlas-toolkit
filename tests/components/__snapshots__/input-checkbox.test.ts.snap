// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCheckbox component match snapshot 1`] = `
<input id="Input"
       type="checkbox"
       part="input"
>
<label for="Input"
       part="label-wrapper"
>
  <slot name="label"
        part="label"
  >
  </slot>
  <slot name="required"
        part="required"
  >
    *
  </slot>
</label>
<slot name="helperText"
      part="helper-text"
>
</slot>
<slot name="validationText"
      part="validation-text"
>
</slot>
`;
